contracts/SpokePool.sol

```diff
@@ -146,20 +174,73 @@ abstract contract SpokePool is Testable, Lockable, MultiCaller {
+    function fillRelay(
         address sender,
-        uint256 amount,
         address recipient,
-        uint256 relayerFee,
-        uint256 realizedLpFee
-    ) public {}
-
-    function fillRelay(
-        uint256 relayId,
-        uint256 fillAmount,
+        address destinationToken,
+        uint64 realizedLpFeePct,
+        uint64 relayerFeePct,
+        uint64 depositId,
+        uint256 originChainId,
+        uint256 totalRelayAmount,
+        uint256 maxTokensToSend,
         uint256 repaymentChain
-    ) public {}
+    ) public {
+        // We limit the relay fees to prevent the user spending all their funds on fees.
+        require(
+            relayerFeePct <= 0.5e18 && realizedLpFeePct <= 0.5e18 && (relayerFeePct + realizedLpFeePct) < 1e18,
+            "invalid fees"
+        );
+
+        // Each relay attempt is mapped to the hash of data uniquely identifying it, which includes the deposit data
+        // such as the origin chain ID and the deposit ID, and the data in a relay attempt such as who the recipient
+        // is, which chain and currency the recipient wants to receive funds on, and the relay fees.
+        RelayData memory relayData = RelayData({
+            sender: sender,
+            recipient: recipient,
+            destinationToken: destinationToken,
+            realizedLpFeePct: realizedLpFeePct,
+            relayerFeePct: relayerFeePct,
+            depositId: depositId,
+            originChainId: originChainId,
+            relayAmount: totalRelayAmount
+        });
+        bytes32 relayHash = _getRelayHash(relayData);
+
+        // Check that the caller is filling a non zero amount of the relay and the relay has not already been completely
+        // filled. Note that the `relays` mapping will point to the amount filled so far for a particular `relayHash`,
+        // so this will start at 0 and increment with each fill.
+        require(
+            maxTokensToSend > 0 && relayFills[relayHash] < totalRelayAmount,
+            "Cannot send 0, or relay filled"
+        );
+
+        // Compute the equivalent amount to be sent by the relayer before fees have been taken out. This is the amount
+        // that we'll add to the `relayFills` counter, and we do this math here in the contract for the user's
+        // convenience so that they don't have to do this math before calling this function. The user can simply
+        // pass in `maxTokensToSend` and assume that the contract will pull exactly that amount of tokens (or revert).
+        uint256 fillAmountPreFees = 1e18 * maxTokensToSend / (1e18 - (realizedLpFeePct + relayerFeePct));
+
+        // If user's specified max amount to send is greater than the amount of the relay remaining pre-fees, 
+        // we'll pull exactly enough tokens to complete the relay.
+        uint256 amountToSend;
+        if (totalRelayAmount - relayFills[relayHash] < fillAmountPreFees) {
+            amountToSend = totalRelayAmount - relayFills[relayHash];
```

Comment:
Won't this cause them to send the amount _including fees_. Don't we need to multiply this by `(1e18 - (realizedLpFeePct + relayerFeePct)) / 1e18`?

Reply:
Ah yes good call, fixed